ghcide/src/Development/IDE/Core/Debouncer.hs

```diff
@@ -31,28 +31,26 @@ newtype Debouncer k = Debouncer { registerEvent :: Seconds -> k -> IO () -> IO (
 
 -- | Debouncer used in the IDE that delays events as expected.
 newAsyncDebouncer :: (Eq k, Hashable k) => IO (Debouncer k)
-newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> newVar Map.empty
+newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> STM.newIO
 
 -- | Register an event that will fire after the given delay if no other event
 -- for the same key gets registered until then.
---
--- If there is a pending event for the same key, the pending event will be killed.
--- Events are run unmasked so it is up to the user of `registerEvent`
--- to mask if required.
-asyncRegisterEvent :: (Eq k, Hashable k) => Var (HashMap k (Async ())) -> Seconds -> k -> IO () -> IO ()
-asyncRegisterEvent d 0 k fire = do
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Nothing)) k m
-        return (m', cancel)
-    fire
-asyncRegisterEvent d delay k fire = mask_ $ do
-    a <- asyncWithUnmask $ \unmask -> unmask $ do
-        sleep delay
-        fire
-        modifyVar_ d (evaluate . Map.delete k)
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Just a)) k m
-        return (m', cancel)
+asyncRegisterEvent :: (Eq k, Hashable k) => STM.Map k (TVar (Seconds, IO())) -> Seconds -> k -> IO () -> IO ()
+asyncRegisterEvent d delay k fire = join $ atomicallyNamed "debouncer - register" $ do
+    prev <- STM.lookup k d
+    case prev of
+        Just v -> writeTVar v (delay, fire) >> return (pure ())
+        Nothing
+          | delay == 0 -> return fire
+          | otherwise  -> do
+            var <- newTVar (delay, fire)
+            STM.insert var k d
+            return $ void $ async $
+                join $ atomicallyNamed "debouncer - sleep" $ do
+                    (s,act) <- readTVar var
```

Comment:
Do we actually need the inner TVar here? What if the async thread just did an STM map lookup to get the value? Wouldn't that have the right properties already?

Reply:
Not exactly, as the transaction would range over all the contents of the STM Map (or rather, all the contents of the branch that holds this key)

Reply:
Wait, but it already does that because it has to delete the key at the end!

Reply:
That's no good. The delete needs to happen in a separate transaction. But then there is a tiny race condition window. Hmm

Reply:
Okay, so one option would be to just accept that we're locking a bit of the map, and write some simpler code that just looks up the value in the map.

Another perhaps ugly idea: store a `TMVar` instead of a `TVar` in the map. Then the async task can:
1. In a transaction: try to look at (and empty) the `TMVar`, and then do the action
2. In a transaction: look at the `TMVar`, if it's empty, then remove it from the map (otherwise someone has put something else in there, so we want to leave it anyway).

That is, we use a `TMVar` as a one element queue, and the async thread first draws from the queue atomically and executes an action, and then optionally tries to garbage-collect the queue so as to keep the map small.

Reply:
So then there's a race window between 1 and 2, but it's benign, because all that can happen is that someone has put something in the queue again, in which case we can just not garbage-collect it.

Reply:
Ok, I have gone with a `TVar (Maybe ..)` and simplified your solution a bit by never deleting from the map

Reply:
That makes me a little nervous - isn't that a space leak?

Reply:
Technically yes, but a bounded one. Only in projects with infinite files would this be a worry.

Reply:
Worth a comment.